pss/crypto/crypto.go

```diff
@@ -0,0 +1,656 @@
+	signatureFlag           = byte(4)
+	aesKeyLength            = 32 // in bytes
+
+	defaultPaddingByteSize = 16
+)
+
+type WrapParams struct {
+	Src    *ecdsa.PrivateKey
+	Dst    *ecdsa.PublicKey
+	KeySym []byte
+}
+
+// Contains a successfully decrypted message prior to parsing and validating
+type ReceivedMessage interface {
+	ValidateAndParse() bool
+	GetPayload() []byte
+	GetSrc() *ecdsa.PublicKey
+}
+
+// Crypto contains methods from MessageCrypto, CryptoBackend and CryptoUtils
+type Crypto interface {
+	MessageCrypto
+	CryptoBackend
+	CryptoUtils
+}
+
+// MessageCrypto contains methods for wrapping(encrypting) and unwrapping(decrypting) messages
+type MessageCrypto interface {
+	WrapMessage(msgData []byte, params *WrapParams) (data []byte, err error)
+	UnWrapSymmetric(encryptedData, key []byte) (ReceivedMessage, error)
+	UnWrapAsymmetric(encryptedData []byte, key *ecdsa.PrivateKey) (ReceivedMessage, error)
+}
+
+// CryptoBackend contains key manipulation methods
+type CryptoBackend interface {
+
+	// Key store functions
+	GetSymKey(id string) ([]byte, error)
+	GenerateSymKey() (string, error)
+	AddSymKeyDirect(bytes []byte) (string, error)
+
+	// Key conversion
+	FromECDSAPub(pub *ecdsa.PublicKey) []byte
+
+	// Key serialization
+	UnmarshalPubkey(pub []byte) (*ecdsa.PublicKey, error)
+	CompressPubkey(pubkey *ecdsa.PublicKey) []byte
+}
+
+// CryptoUtils contains utility methods for tests
+type CryptoUtils interface {
+	GenerateKey() (*ecdsa.PrivateKey, error)
+	NewKeyPair(ctx context.Context) (string, error)
+	GetPrivateKey(id string) (*ecdsa.PrivateKey, error)
+}
+
+var (
+	errInvalidPubkey               = errors.New("invalid public key provided for asymmetric encryption")
+	errInvalidSymkey               = errors.New("invalid key provided for symmetric encryption")
+	errMissingSaltOrInvalidPayload = errors.New("missing salt or invalid payload in symmetric message")
+	errSecureRandomData            = errors.New("failed to generate secure random data")
+	errNeitherSymNorAsymKeysProv   = errors.New("unable to encrypt the message: neither symmetric nor asymmetric key provided")
+)
+
+type defaultCryptoBackend struct {
+	privateKeys map[string]*ecdsa.PrivateKey // Private key storage
+	symKeys     map[string][]byte            // Symmetric key storage
+	keyMu       sync.RWMutex                 // Mutex associated with key storages
+}
+
+// receivedMessage represents a data packet to be received
+// and successfully decrypted.
+type receivedMessage struct {
+	Payload   []byte // Parsed and validated message content
+	Raw       []byte // Unparsed but decrypted message content
+	Signature []byte // Signature from the sender
+	Salt      []byte // Salt used for validation
+	Padding   []byte // Padding applied
+
+	Src *ecdsa.PublicKey // Source public key used for signing the message
+	Dst *ecdsa.PublicKey // Destination public key used for encrypting the msessage asymmetrically
+
+	crypto *defaultCryptoBackend
```

Comment:
In order to call ValidateAndParse, yes, I need that or I will need to pass crypto as an argument to GetPayload. But if we have several cryptos at the same time decrypting different messages, I think adding crypto in the message makes the API more robust because the interface-user could not use the wrong crypto implementation.